src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +619,263 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
 }
 
 
+/*
+ * RecursivelyPlanRecurringTupleOuterJoins descends into a join tree and
+ * recursively plans all non-recurring (i.e., distributed) rels that that
+ * participate in an outer join expression together with a recurring rel,
+ * such as <distributed> in "<recurring> LEFT JOIN <distributed>", i.e.,
+ * where the recurring rel causes returning recurring tuples from the worker
+ * nodes.
+ *
+ * Returns true if given node is recurring.
+ */
+static bool
+RecursivelyPlanRecurringTupleOuterJoins(Node *node, Query *query,
+										RecursivePlanningContext *recursivePlanningContext)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+	else if (IsA(node, RangeTblRef))
+	{
+		return IsRTERefRecurring((RangeTblRef *) node, query);
+	}
+	else if (IsA(node, FromExpr))
+	{
+		FromExpr *fromExpr = (FromExpr *) node;
+		ListCell *fromExprCell;
+
+		bool allRecurring = true;
+
+		/* search for join trees in each FROM element */
+		foreach(fromExprCell, fromExpr->fromlist)
+		{
+			Node *fromElement = (Node *) lfirst(fromExprCell);
+
+			if (!RecursivelyPlanRecurringTupleOuterJoins(fromElement, query,
+														 recursivePlanningContext))
+			{
+				allRecurring = false;
+			}
+		}
+
+		return allRecurring;
+	}
+	else if (IsA(node, JoinExpr))
+	{
+		JoinExpr *joinExpr = (JoinExpr *) node;
+
+		Node *leftNode = joinExpr->larg;
+		Node *rightNode = joinExpr->rarg;
+
+		/*
+		 * There may be recursively plannable outer joins deeper in the join tree.
+		 *
+		 * We first handle the sub join trees and then the top level one since the
+		 * top level join expression might not require recursive planning after
+		 * handling the sub join trees.
+		 */
+		bool leftNodeRecurs = RecursivelyPlanRecurringTupleOuterJoins(leftNode, query,
+																	  recursivePlanningContext);
+		bool rightNodeRecurs = RecursivelyPlanRecurringTupleOuterJoins(rightNode, query,
+																	   recursivePlanningContext);
+		switch (joinExpr->jointype)
```

Comment:
I guess we are protected by semijoins with other checks. I could not find a case where we end up processing semi-joins here:

```SQL
SELECT *
FROM a
WHERE EXISTS
    (SELECT 1
     FROM b
     WHERE b.i = a.i );

ERROR:  correlated subqueries are not supported when the FROM clause contains a reference table
```

And, anti-joins are represented as OUTER JOINs. But then, why `IS_OUTER_JOIN ` has `JOIN_ANTI ` as a check? Can you dig down a bit?
```SQL
EXPLAIN
SELECT count(*)
FROM a
LEFT JOIN b ON (a.i = b.i)
WHERE b.i IS NULL;
```

For both cases, would be nice to document why we skip.


Also, as a reminder, you can find interesting queries from here: https://github.com/postgres/postgres/blob/51dfaa0b010c1efba8a4137cfccd5fbb7f5d4afd/src/test/regress/expected/join.out

Reply:
For me IS_OUTER_JOIN() is a context-agnostic macro that doesn't know about whether the planner converts anti joins to other outer join types but only knows what is an outer join basically.